src/main/scala/com/github/gvolpe/fs2/Once.scala

```diff
@@ -26,28 +21,29 @@ object OnceApp extends Once[IO]
   * 10 seconds and you can know for sure that the process completing the promise is going to
   * be the first one.
   * */
-class Once[F[_]: Effect] extends StreamApp[F] {
+object OnceApp extends IOApp {
+  import cats.syntax.all._
 
-  override def stream(args: List[String], requestShutdown: F[Unit]): fs2.Stream[F, ExitCode] =
-    Scheduler(corePoolSize = 4).flatMap { implicit scheduler =>
-      for {
-        p <- Stream.eval(async.promise[F, Int])
-        e <- new ConcurrentCompletion[F](p).start
-      } yield e
-    }
+  def stream[F[_]: ConcurrentEffect]: fs2.Stream[F, Unit] =
+    for {
+      p <- Stream.eval(Deferred[F, Int])
+      e <- new ConcurrentCompletion[F](p).start
+    } yield e
 
+  override def run(args: List[String]): IO[ExitCode] =
+    stream[IO].compile.drain.as(ExitCode.Success)
 }
 
-class ConcurrentCompletion[F[_]](p: Promise[F, Int])(implicit F: Effect[F]) {
+class ConcurrentCompletion[F[_]](p: Deferred[F, Int])(implicit F: Concurrent[F]) {
 
   private def attemptPromiseCompletion(n: Int): Stream[F, Unit] =
     Stream.eval(p.complete(n)).attempt.drain
 
-  def start: Stream[F, ExitCode] =
+  def start: Stream[F, Unit] =
```

Comment:
Can be defined without streams, maybe something like (haven't tried though):

```scala
val exec: F[Unit] =
  List(
    p.complete(1), 
    p.complete(2), 
    p.get.flatMap(n => putStrLn(s"Result: $n"))
  ).parTraverse(_.attempt).void
```

Reply:
👍
done